File: Semester 1/hometask_7/Source_7.2/avlTree.cpp
#include "avlTree.h"
#include <iostream>

Node* createNode(int value) {
	Node *newNode = new Node;
	
	newNode->key = value;
	newNode->height = 0;
	newNode->left = nullptr;
	newNode->right = nullptr;

	return newNode;
}

int height(Node *node) {
	return node ? node->height : 0;
}

int balanceFactor(Node *node) {
	return height(node->right) - height(node->left);
}

void updateHeight(Node *node) {
	int heightLeft = height(node->left);
	int heightRight = height(node->right);
	node->height = ((heightLeft > heightRight) ? heightLeft : heightRight) + 1;
}

Node* rotateRight(Node *root) {
	Node* pivot = root->left;
	root->left = pivot->right;
	pivot->right = root;

	updateHeight(root);
	updateHeight(pivot);
	
	return pivot;
}

Node* rotateLeft(Node *root) {
	Node* pivot = root->right;
	root->right = pivot->left;
	pivot->left = root;

	updateHeight(root);
	updateHeight(pivot);

	return pivot;
}

Node* balance(Node* p) {
	updateHeight(p);

	if (balanceFactor(p) == 2) {
		if (balanceFactor(p->right) < 0)
			p->right = rotateRight(p->right);

		return rotateLeft(p);
	}

	if (balanceFactor(p) == -2) 	{
		if (balanceFactor(p->left) > 0)
			p->left = rotateLeft(p->left);

		return rotateRight(p);
	}

	return p;
}

void add(int value, Node *&p) {
	if (p == nullptr) {
		p = createNode(value);
		p = balance(p);
		return;
	}

	if (value < p->key)
		add(value, p->left);
	else
		add(value, p->right);

	p = balance(p);
}

void add(int value, AvlTree &tree) {
	add(value, tree.root);
}

Node* findmin(Node* p) {
	return p->left ? findmin(p->left) : p;
}

Node* removemin(Node* p) {
	if (p->left == 0)
		return p->right;
	p->left = removemin(p->left);
	return balance(p);
}

Node* remove (int k, Node* p) {
	if (p == nullptr) 
		return nullptr;
	if (k < p->key)
		p->left = remove(k, p->left);
	else {
		if (k > p->key)
			p->right = remove(k, p->right);
		else {
			Node* left = p->left;
			Node* right = p->right;
			delete p;
			if (!right) 
				return left;
			Node* min = findmin(right);
			min->right = removemin(right);
			min->left = left;
			return balance(min);
		}
	}
	return balance(p);
}

void remove(int value, AvlTree &tree) {
	tree.root = remove(value, tree.root);
}

void outByLevel(Node *p) {
	if (p != nullptr) {
		std::cout << " (" << p->key;
		outByLevel(p->left);
		std::cout << "";
		outByLevel(p->right);
		std::cout << ")";
	}
	else
		std::cout << " NULL";
}

void abcOut(AvlTree &tree) {
	outByLevel(tree.root);
	std::cout << std::endl;
}

bool isContains(int value, Node *p) {
	if (p == nullptr)
		return false;

	if (p->key == value)
		return true;

	if (value < p->key)
		return isContains(value, p->left);
	else
		return isContains(value, p->right);
}

bool isContains(int value, AvlTree &tree) {
	return isContains(value, tree.root);
}

bool isEmpty(AvlTree &tree) {
	return tree.root == nullptr;
}

void incOut(Node *p) {
	if (p == nullptr)
		return;

	incOut(p->left);
	std::cout << " " << p->key << " ";
	incOut(p->right);
}

void increasOut(AvlTree &tree) {
	incOut(tree.root);
	std::cout << std::endl;
}

void decOut(Node *p) {
	if (p == nullptr)
		return;

	decOut(p->right);
	std::cout << " " << p->key << " ";
	decOut(p->left);
}

void decreasOut(AvlTree &tree) {
	decOut(tree.root);
	std::cout << std::endl;
}

void clear(Node *p) {
	if (p == nullptr)
		return;

	clear(p->left);
	clear(p->right);

	delete p;
}

void clear(AvlTree &tree) {
	clear(tree.root);
}

File: semester_2/hometask_4/Source_4.2/src/main/java/group144/khodko/AVLTree.java
package group144.khodko;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * AVL tree custom implementation using {@link Collection} interface
 *
 * @param <T> the type of objects that can be stored in AVL tree
 * @author a1rick
 */
public class AVLTree<T extends Comparable<T>> implements Collection<T> {

    private Node<T> root;

    private int size = 0;

    private static String printNode(Node node) {
        return node == null ? "" : String.format("(%s left:%s right:%s)", node.key.toString(), printNode(node.left), printNode(node.right));
    }

    private static <K extends Comparable<K>> int height(Node<K> node) {
        return node != null ? node.height : 0;
    }

    private static <K extends Comparable<K>> int balanceFactor(Node<K> node) {
        return height(node.right) - height(node.left);
    }

    private static <K extends Comparable<K>> void fixHeight(Node<K> node) {
        node.height = Math.max(height(node.left), height(node.right)) + 1;
    }

    private static <K extends Comparable<K>> Node<K> rotateRight(Node<K> node) {
        Node<K> left = node.left;
        node.left = left.right;
        left.right = node;
        fixHeight(node);
        fixHeight(left);
        return left;
    }

    private static <K extends Comparable<K>> Node<K> inOrderSuccessor(Node<K> node, Node<K> root) {
        if (node.right != null) return findMin(node.right);
        Node<K> successor = null;
        while (root != null) {
            int compare = node.key.compareTo(root.key);
            if (compare < 0) {
                successor = root;
                root = root.left;
            } else if (compare > 0) root = root.right;
            else break;
        }
        return successor;
    }

    private static <K extends Comparable<K>> Node<K> rotateLeft(Node<K> node) {
        Node<K> right = node.right;
        node.right = right.left;
        right.left = node;
        fixHeight(node);
        fixHeight(right);
        return right;
    }

    private static <K extends Comparable<K>> Node<K> balance(Node<K> node) {
        fixHeight(node);
        int balanceFactor = balanceFactor(node);
        if (balanceFactor == 2) {
            if (balanceFactor(node.right) < 0) {
                node.right = rotateRight(node.right);
            }
            return rotateLeft(node);
        } else if (balanceFactor == -2) {
            if (balanceFactor(node.left) > 0) {
                node.left = rotateLeft(node.left);
            }
            return rotateRight(node);
        }
        return node;
    }

    private static <K extends Comparable<K>> Node<K> insert(Node<K> node, K key) {
        if (node == null)
            return new Node<>(key);
        else if (key.compareTo(node.key) < 0) {
            node.left = insert(node.left, key);
        } else {
            node.right = insert(node.right, key);
        }
        return balance(node);
    }

    private static <K extends Comparable<K>> Node<K> findMin(Node<K> node) {
        return node.left == null ? node : findMin(node.left);
    }

    private static <K extends Comparable<K>> Node<K> removeMin(Node<K> node) {
        if (node.left == null)
            return node.right;
        node.left = removeMin(node.left);
        return balance(node);
    }

    private static <K extends Comparable<K>> Node<K> remove(Node<K> node, Comparable<K> key) {
        if (node == null) return null;
        else {
            int compare = key.compareTo(node.key);
            if (compare < 0) {
                node.left = remove(node.left, key);
            } else if (compare > 0) {
                node.right = remove(node.right, key);
            } else {
                Node<K> left = node.left;
                Node<K> right = node.right;
                if (right == null)
                    return left;
                Node<K> min = findMin(right);
                min.right = removeMin(right);
                min.left = left;
                return balance(min);
            }
        }
        return balance(node);
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    @SuppressWarnings("unchecked")
    public boolean contains(Object o) {
        if (o == null || !(o instanceof Comparable)) return false;
        Comparable<T> comparable = (Comparable<T>) o;
        Node<T> temp = root;
        while (temp != null) {
            int compare = comparable.compareTo(temp.key);
            if (compare == 0) return true;
            temp = compare < 0 ? temp.left : temp.right;
        }
        return false;
    }

    @Override
    public Iterator<T> iterator() {
        return new AVLTreeIterator<>(root, root != null ? findMin(root) : null);
    }

    @Override
    public Object[] toArray() {
        Object[] array = new Object[size];
        int i = 0;
        for (T t : this) {
            array[i++] = t;
        }
        return array;
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T1> T1[] toArray(T1[] a) {
        if (a.length < size) return (T1[]) toArray();
        int i = 0;
        for (T t : this) {
            a[i++] = (T1) t;
        }
        for (int j = i; j < a.length; j++) {
            a[j] = null;
        }
        return a;
    }

    @Override
    public boolean add(T t) {
        size++;
        root = insert(root, t);
        return true;
    }

    @Override
    public boolean remove(Object o) {
        if (o == null || !contains(o)) return false;
        size--;
        root = remove(root, (Comparable<T>) o);
        return true;
    }

    @Override
    public boolean containsAll(Collection<?> c) {
        for (Object o : c) {
            if (!contains(o)) return false;
        }
        return true;
    }

    @Override
    public boolean addAll(Collection<? extends T> c) {
        for (T t : c) {
            add(t);
        }
        return true;
    }

    @Override
    public boolean removeAll(Collection<?> c) {
        boolean removed = false;
        for (Object t : c) {
            removed |= remove(t);
        }
        return removed;
    }

    @Override
    public boolean retainAll(Collection<?> c) {
        boolean removed = false;
        for (T t : this) {
            if (!c.contains(t)) removed |= remove(t);
        }
        return removed;
    }

    @Override
    public void clear() {
        root = null;
        size = 0;
    }

    @Override
    public String toString() {
        return new ArrayList<>(this).toString();
    }

    String printTree() {
        return printNode(root);
    }

    private static final class Node<K> {
        private K key;
        private int height = 1;
        private Node<K> left = null;
        private Node<K> right = null;

        private Node(K key) {
            this.key = key;
        }
    }

    private static final class AVLTreeIterator<E extends Comparable<E>> implements Iterator<E> {

        private final Node<E> root;
        private Node<E> current;

        private AVLTreeIterator(Node<E> root, Node<E> min) {
            this.root = root;
            this.current = min;
        }

        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public E next() {
            if (current == null) throw new NoSuchElementException("No elements found");
            Node<E> prev = this.current;
            current = inOrderSuccessor(this.current, root);
            return prev.key;
        }
    }
}